elastic/kibana
# [Search Sessions] Improve session restoration back button

## Summary

Closes https://github.com/elastic/kibana/issues/82419

Improves back button behavior for search session restoration state
Discover and dashboard because this is only apps we have session restoration flow for now

### How to test

Full flow will be testable after https://github.com/elastic/kibana/pull/81707 is merged, but for now:

1. Add a fake `searchSessionId` to discover or dashboard URL. Load it. The session should fail to restore in this case. (you can also use a real `seaerchSessionId` using inspector)
2. Change the state that would trigger a new search. e.g. change time
3. Press back -> should return you in the previous state with `searchSessionId` in the URL

https://user-images.githubusercontent.com/7784120/103906732-9ebd7b80-5100-11eb-8117-2da841496da5.mov



Both for Discover and Dashboard there are different main cases with this navigation that are different from one another and expected to work: 

1. In case navigating away from a restored session together with another state change, like changing time 
2. In case navigating away from a restored session without another state change, like when hitting "refresh" 
3. another case that should work: directly changing the `searchSessionId` in the browser address bar

Discover has a very specific code path when changing index patterns. This is another code path that is worth testing and it should work.


### Implemintation 

I fully exposed underlying `kbnUrlControls` on `kbnUrlStorage`. Dashboard needs them to remove `searchSessionId` during pending state sync update. 

~There is a known edge case that if you open a Discover or Dashboard URL and then change only `searchSessionId` then nothing would happen.~
~That is because there is no listener that listens specifically to that query param. Such change is happening in a real-world scenario in case the "refresh" button was hit.~

- [x] ~@Dosant: Todo: try fixing this, see how difficult this is...~ This if fixed but made implementation more complicated :( 


**In Dashboard** I had to add a denounce on updating dashboard container input to cover the following edge case: 

1. Restoring a session -> navigating away with other state change, for example, with time range change
2. time is changed in the URL. `searchSessionId` is removed from the URL.
3. Hit back 
4. both `connectToGlobalQueryState` (time) and `searchSessionIdQuery$` will ask for dashboard container update. to avoid double fetch in this cases I added debounce.  



### Notes

* Please note when testing in Discover: https://github.com/elastic/kibana/pull/87635#issuecomment-758784752

### Checklist

Delete any items that are not applicable to this PR.


### For maintainers
